include/wil/com_apartment_variable.h

```
@@ -0,0 +1,377 @@
+                }
+                else
+                {
+                    struct ApartmentObserver : public winrt::implements<ApartmentObserver, IApartmentShutdown>
+                    {
+                        void STDMETHODCALLTYPE OnUninitialize(unsigned long long apartmentId) noexcept override
+                        {
+                            // This code runs at apartment rundown so be careful to avoid deadlocks by
+                            // extracting the variables under the lock then release them outside.
+                            auto variables = [apartmentId]()
+                            {
+                                auto lock = winrt::slim_lock_guard(s_lock);
+                                return s_apartmentStorage.extract(apartmentId);
+                            }();
+                        }
+                    };
+
+                    test_hook::RegisterForApartmentShutdown(winrt::make<ApartmentObserver>().get());
+
+                    auto variables = s_apartmentStorage.insert({ apartmentId,
+                        { { this, creator() } } }).first;
+                    return variables->second.begin()->second; // new map entry is first
+                }
+            }
+
+            // get pointer to current value or nullptr if no value has been set
+            std::any* get_if()
+            {
+                auto lock = winrt::slim_lock_guard(s_lock);
+
+                auto storage = s_apartmentStorage.find(test_hook::GetApartmentId());
+                if (storage != s_apartmentStorage.end())
+                {
+                    auto& variables = storage->second;
+                    auto variable = variables.find(this);
+                    if (variable != variables.end())
+                    {
+                        return &(variable->second);
+                    }
+                }
+                return nullptr;
+            }
+
+            // replace or create the current value

```

Comment says it can create, but code fails fast if value does not exist.